### PR TITLE
Add Kerberos Authentication

### PR DESCRIPTION
**Pull request by leefine02**

### CODE REVIEW

File: hashicorp-vault-pam/VaultAPI.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+                logger.LogDebug($"PAM Provider {name} - requesting client auth token from kerberos authentication request at {host}");
+                var kerberosResponse = client.PostAsync($"{host}{authEndpoint}", null).GetAwaiter().GetResult();
+                logger.LogTrace($"PAM Provider {name} - received response to kerberos auth request");
+
+                kerberosResponse.EnsureSuccessStatusCode();
+                var kerberosContent = kerberosResponse.Content.ReadAsStringAsync().GetAwaiter().GetResult();
+                JObject kerb = JObject.Parse(kerberosContent);
+
+                logger.LogDebug($"PAM Provider {name} - returning secret from vault");
```
Comment (doebrowsk) [inline hashicorp-vault-pam/VaultAPI.cs:78]: this log message looks incorrect

File: hashicorp-vault-pam/VaultAPI.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
+﻿// Copyright 2022 Keyfactor
```
Comment (doebrowsk) [inline hashicorp-vault-pam/VaultAPI.cs:1]: copyright is out of date (update in any files updated with this release)

File: hashicorp-vault-pam/manifesTokenAuth.json
Comment (doebrowsk) [inline hashicorp-vault-pam/manifesTokenAuth.json:1]: file name has a typo

**Review by doebrowsk — CHANGES_REQUESTED** [review]
Besides the comments on the commit files, I feel like the commit history looks kind of incorrect here. All of the commits have the message as "changes" which is not super helpful, but I bet is probably difficult to retroactively change at this point. At the least I think the changes should get squashed into one commit on merge and given a good commit message. 

More importantly, the user of most of the commits is listed as 'kfadmin' which I think needs to be changed before this can be merged. This is most likely because of local git settings on the machine that made the commits, where this would need to be updated for the future.

File: hashicorp-vault-pam/VaultAPI.cs (hunk 2 above)
Comment (leefine02) [reply hashicorp-vault-pam/VaultAPI.cs:1]: Are we supposed to just update these to the current year whenever we modify something?

File: hashicorp-vault-pam/VaultAPI.cs (hunk 1 above)
Comment (leefine02) [reply hashicorp-vault-pam/VaultAPI.cs:78]: Fixed

**Review by RexWheeler — APPROVED** [review]
Looking at the readme, I think it is reasonable. We do call that the kerberos setup is out of scope. I wonder if we should be more explicit that it is also not something that our support team generally assists with. Do we have a technical test that we could publish that would test that kerberos is set up properly - similar to how we have some PowerShell scripts that make the same API calls as our other integrations do, which can be used to prove incorrect configurations? Ideally we would be able to have a customer run a couple commands or a script to verify that a usable token can be created (and do this from outside the context of our PAM module). I don't think we need to hold the release or review of this pending such a script, but it would be a good thing to have.

File: hashicorp-vault-pam/VaultAPI.cs (hunk 2 above)
Comment (fiddlermikey) [reply hashicorp-vault-pam/VaultAPI.cs]: Are the dates to be updated in all project files or just the ones being modified? There should be a tool to update the date if all files need updating.

Comment (leefine02) [reply hashicorp-vault-pam/VaultAPI.cs]: > Are the dates to be updated in all project files or just the ones being modified? There should be a tool to update the date if all files need updating.

MIkey, I just updated all 3 code files.  It's a small repo.

**Comment by leefine02** [comment]
> Looking at the readme, I think it is reasonable. We do call that the kerberos setup is out of scope. I wonder if we should be more explicit that it is also not something that our support team generally assists with. Do we have a technical test that we could publish that would test that kerberos is set up properly - similar to how we have some PowerShell scripts that make the same API calls as our other integrations do, which can be used to prove incorrect configurations? Ideally we would be able to have a customer run a couple commands or a script to verify that a usable token can be created (and do this from outside the context of our PAM module). I don't think we need to hold the release or review of this pending such a script, but it would be a good thing to have.

Rex, I don't have a script to test the Kerberos set up, but the link I have at the end of the README has a CLI version of the Login API.  That is what I used to test before moving on to the integration itself.

**Comment by RexWheeler** [comment]
> > Looking at the readme, I think it is reasonable. We do call that the kerberos setup is out of scope. I wonder if we should be more explicit that it is also not something that our support team generally assists with. Do we have a technical test that we could publish that would test that kerberos is set up properly - similar to how we have some PowerShell scripts that make the same API calls as our other integrations do, which can be used to prove incorrect configurations? Ideally we would be able to have a customer run a couple commands or a script to verify that a usable token can be created (and do this from outside the context of our PAM module). I don't think we need to hold the release or review of this pending such a script, but it would be a good thing to have.
> 
> Rex, I don't have a script to test the Kerberos set up, but the link I have at the end of the README has a CLI version of the Login API. That is what I used to test before moving on to the integration itself.

It wasn't clear to me what CLI command you are referring to. Perhaps in our readme we should have some text like: "Ensure that the following command (mentioned in the Vault link) completes successfully: [whatever the command is]. This command can be used as a smoke test to ensure that kerberos authentication is properly set up in your environment with Vault."

**Comment by leefine02** [comment]
> > > Looking at the readme, I think it is reasonable. We do call that the kerberos setup is out of scope. I wonder if we should be more explicit that it is also not something that our support team generally assists with. Do we have a technical test that we could publish that would test that kerberos is set up properly - similar to how we have some PowerShell scripts that make the same API calls as our other integrations do, which can be used to prove incorrect configurations? Ideally we would be able to have a customer run a couple commands or a script to verify that a usable token can be created (and do this from outside the context of our PAM module). I don't think we need to hold the release or review of this pending such a script, but it would be a good thing to have.
> > 
> > 
> > Rex, I don't have a script to test the Kerberos set up, but the link I have at the end of the README has a CLI version of the Login API. That is what I used to test before moving on to the integration itself.
> 
> It wasn't clear to me what CLI command you are referring to. Perhaps in our readme we should have some text like: "Ensure that the following command (mentioned in the Vault link) completes successfully: [whatever the command is]. This command can be used as a smoke test to ensure that kerberos authentication is properly set up in your environment with Vault."

Rex, I added a step 11 with the Login command.

**Pull request merged**